### ISSUE TITLE
[feature] cmake toolchains should be generator agnostic

### ISSUE THREAD

**Issue by ohanar** [conan-io/conan#7908]
We regularly use a few different cmake generators for development, even when targeting a single platform. Because `conan_toolchain.cmake` has different content depending on the generator conan is targeting, right now our developers have to regularly change their `CONAN_CMAKE_GENERATOR` environment variable when switching generators, which ends up being a fiddly process that is frequently forgotten, and leads to cryptic errors.

Additionally, for recipes that hard-code a specific generator, a developer of a conan package has to both pass the generator to the `CMakeToolchain` as well as the `CMake` build helper. This is again easy to forget, especially when the cmake toolchain doesn't need any other special configuration (i.e. `toolchain = "cmake"` fails to work as a developer might expect).

Instead, the `conan_toolchain.cmake` file's contents should not depend upon the generator. Where generator specific content is required, that can be checked at cmake runtime by examining the `CMAKE_GENERATOR` variable, so for instance the Visual Studio toolset could be set after checking that `CMAKE_GENERATOR` is a Visual Studio project, or similar for passing the `/MP` flag.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @ohanar 

yes, the configuration and variability of the toolchains, is one of the things that we still need to figure out. Not only what the toolchain does, but also, I feel that configuring it via env-vars is probably not the best approach.

I agree, let's start trying to make the toolchain generic conditioned on the generator.

**Comment by mpusz** [comment]
Well, I am not sure if this is correct to not depend on a generator. For example, it seems that the following is being set for the Visual Studio compiler:

```cmake
set(CMAKE_GENERATOR_PLATFORM "x64" CACHE STRING "" FORCE)
```

However, this breaks all the generators beside "Visual Studio" (i.e. Ninja):

https://github.com/mpusz/units/runs/2148187500?check_suite_focus=true

Which effectively makes it impossible to use Visual Studio compiler with `CMakeToolchain` when someone does not want to generate `*.sln` files 😞

**Comment by mpusz** [comment]
I believe that the above should be set only if one of "Visual Studio" CMake generators is being used in CMake.

**Comment by ohanar** [comment]
@mpusz Being generator agnostic doesn't mean that they can't handle generator specific settings, e.g.
```cmake
if ( "${CMAKE_GENERATOR}" MATCHES "^Visual Studio" )
  set ( CMAKE_GENERATOR_PLATFORM "x64" CACHE STRING "" FORCE )
endif()
```

**Comment by ohanar** [comment]
Oh, I think I initially misread your comment. Yes, this is exactly what this issue is about. Right now you can control it with either the environment variable `CONAN_CMAKE_GENERATOR` or hardcoding the generator into the recipe.

**Comment by mpusz** [comment]
Or do `conan install ...` and then use any CMake generator with `cmake .. -G ...`. And it fails for most generators when Visual Studio is the target compiler.

**Comment by ohanar** [comment]
@mpusz Are you loading up the appropriate visual studio environment when trying to run CMake? I use MSVC + Ninja + CMake + Conan's CMakeToolchain every day, so I know that they work together so long as you make sure you initialize your environment correctly.

**Comment by mpusz** [comment]
No, in general. As I am aware of how to do it from the command line, I even do not know how to do it in for example inside CLion or VS Code that natively work with the MSVC compiler and break with exactly the same error?

**Comment by ohanar** [comment]
Are you saying that you are able to do it in the command line but not in an IDE? If so, you'll have to consult the IDE to see if it can handle loading up MSVC's environment and potentially how to set extra environment variables if you are also trying to run conan through them. IDE functionality is unrelated to this issue.

**Comment by mpusz** [comment]
I mean that all IDEs handle MSVC well without actually setting VS env and Conan should support those.

**Comment by mpusz** [comment]
BTW, everything was working fine a few Conan versions ago. I do not know in which exactly version Conan started to generate

```cmake
set(CMAKE_GENERATOR_PLATFORM "x64" CACHE STRING "" FORCE)
```

with `CMakeToolchain` but that broke both CLion and VSCode. Only Visual Studio works fine for me.

**Comment by mpusz** [comment]
Do we have any updates on this issue? Please note that the change above broke CLion and VS Code which natively work with MSVC compiler.

**Comment by memsharded** [comment]
For Conan 1.36, a complete re-write of the ``CMakeToolchain`` internals has been done, specially regarding extensibility and possibility of customization by users.

The important PR is https://github.com/conan-io/conan/pull/8749, but https://github.com/conan-io/conan/pull/8815 contains some changes. To summarize:

- The CMAKE_GENERATOR_PLATFORM is only set for Visual/msvc compiler AND "Visual Studio" CMake generators.
- The ``CMakeToolchain`` is the one defining the CMake generator. It stores it in a text file, so the ``CMake()`` at ``build()`` time, doesn't need to apply any logic at all.
- Beware that env-vars configurations have been removed. We will add [conf] alternatives soon. At the moment, it needs to go as CMakeToolchain argument.

With this changes, I think this issue can be considered closed, but please try 1.36 when it is out and report if necessary. Thanks!

**Comment by mpusz** [comment]
BTW, are you aware of [CMake Presets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html). I think they may supersede CMakeToolchain?

**Comment by memsharded** [comment]
> BTW, are you aware of CMake Presets. I think they may supersede CMakeToolchain?

No, they will not. Conan will support CMake 3.15 as a baseline, everything should work with that version. 

Furthermore, it seems that presets don't fully supersede toolchains, it seems so far that there are a bunch of configurations, flags, etc, specially  for cross-building scenarios that seems not implemented or covered by presets, or at least not the intended solution. Will the Android-NDK toolchain for example be converted to presets? Is that possible at all?

**Comment by ohanar** [comment]
Okay, I gave the dev version of 1.36 a try. This doesn't really solve the problem, cmake toolchain files are still generator dependent:
```
> cmake .. -GNinja -DCMAKE_TOOLCHAIN_FILE="conan_toolchain.cmake"
Using Conan toolchain through conan_toolchain.cmake.
-- Conan toolchain: Setting BUILD_SHARED_LIBS= ON
CMake Error at CMakeLists.txt:9 (project):
  Generator

    Ninja

  does not support platform specification, but platform

    x64

  was specified.


-- Configuring incomplete, errors occurred!
See also "D:/dev/engine/build/tmp/CMakeFiles/CMakeOutput.log".
```
Adding a [conf] just brings us back to the previous status quo (where env variable has been replaced with [conf]). I would expect the cmake toolchain files to not depend at all on the generator, so for instance, with MSVC the toolchain file would always include something along the following

```
if("${CMAKE_GENERATOR}" MATCHES "^Visual Studio")
    set(CMAKE_GENERATOR_PLATFORM "x64" CACHE STRING "" FORCE)
endif()
```

I see that the changes also removes the generator from the build step, and rather writes it in the generate step, which I think is a good change, but it doesn't fundamentally fix this issue.

**Comment by memsharded** [comment]
Hi @ohanar 

The problem is that the build needs to be defined at ``generate()`` time. 

Whatever configuration will be used, including the generator needs to be specified there. The toolchain will do more things, besides setting CMAKE_GENERATOR_PLATFORM. For example, Ninja builds need activated the correct Visual Studio environment, with the correct version, in order to find the msvc compiler. The ``CMakeToolchain`` will generate a ``conanvcvars.bat`` file that contains the right ``vcvars`` command that needs to be used, depending on the settings. But this will only happen if the ``Ninja`` generator is specified there, because for ``Visual Studio`` CMake generators, this doesn't need to be defined as CMake will handle that internally. 

It is also expected that using one generator or the other will imply to have different build folder layouts. We are also working to parameterize the layout and achieve a transparent local-flow, create and editable experience. So even in the specific case of CMAKE_GENERATOR_PLATFORM, this could be workarounded with an if() in the cmake code, there is a strong evidence that the generator needs to be defined well in advance. Actually the ``CMakeToolchain()`` will create a ``conanbuild.json`` file to store this information and use it at ``build()`` time automatically with the ``CMake`` helper. 

In summary, it seems that the ``generate()`` and ``CMakeToolchain`` output cannot be generic enough for various different CMake generators.

**Comment by mpusz** [comment]
Please note that the error mentioned by the @ohanar happens not only on a command line (when you can easily run vcvars.bat before CMake) but also in nearly every IDE (CLion, VS Code, ...) where there is no possibility to run any file before configure step. Please note that those IDEs work great with the MSVC compiler without running those bat files at all so they should not be enforced by Conan as well.

**Comment by memsharded** [comment]
> Please note that the error mentioned by the @ohanar happens not only on a command line (when you can easily run vcvars.bat before CMake) but also in nearly every IDE (CLion, VS Code, ...) where there is no possibility to run any file before configure step. Please note that those IDEs work great with the MSVC compiler without running those bat files at all so they should not be enforced by Conan as well.

How does CLion knows which VS version to use? That was defined in the ``conan install -s compiler.version=14`` command. 
Conan will not force to call ``conanvcvars.bat``, if the environment is correctly setup, vcvars has been called by other means (the same version that was used for ``conan install``), then things will work. 

Still, there are other things that need the CMake generator defined at ``generate()`` time and ``CMakeToolchain()``, and cannot be changed freely after the ``conan install``, for example, the project layout. If we don't do this, then there will be many complains later than the local flow is not possible, that it is not the same when built in the cache and when it is built locally, etc. Not to say people upset because it is very difficult to have packages in "editable" mode. If the only Conan command that is run is ``conan install``, then definitely Conan needs to know the generator that is going to be used, it is not possible to change it later and have everything working automagically.

The error above will not happen if "Ninja" is specified either in the ``generate()`` method (``generator="Ninja"`` argument), or in a Conan [conf], like ``conan install .... --conf tools.cmake:generator=Ninja``.

**Comment by julienalofs** [comment]
When I use conan install .... --conf tools.cmake:generator=Ninja, the conan_toolchain.cmake file generated still contains `set(CMAKE_GENERATOR_PLATFORM "x64" CACHE STRING "" FORCE)` that prevents Ninja to work.

Conan version = 1.44.1

**Comment by julienalofs** [comment]
> > Please note that the error mentioned by the @ohanar happens not only on a command line (when you can easily run vcvars.bat before CMake) but also in nearly every IDE (CLion, VS Code, ...) where there is no possibility to run any file before configure step. Please note that those IDEs work great with the MSVC compiler without running those bat files at all so they should not be enforced by Conan as well.
> 
> How does CLion knows which VS version to use? That was defined in the `conan install -s compiler.version=14` command. Conan will not force to call `conanvcvars.bat`, if the environment is correctly setup, vcvars has been called by other means (the same version that was used for `conan install`), then things will work.
> 
> Still, there are other things that need the CMake generator defined at `generate()` time and `CMakeToolchain()`, and cannot be changed freely after the `conan install`, for example, the project layout. If we don't do this, then there will be many complains later than the local flow is not possible, that it is not the same when built in the cache and when it is built locally, etc. Not to say people upset because it is very difficult to have packages in "editable" mode. If the only Conan command that is run is `conan install`, then definitely Conan needs to know the generator that is going to be used, it is not possible to change it later and have everything working automagically.
> 
> The error above will not happen if "Ninja" is specified either in the `generate()` method (`generator="Ninja"` argument), or in a Conan [conf], like `conan install .... --conf tools.cmake:generator=Ninja`.

When I use `conan install .... --conf tools.cmake:generator=Ninja`, the conan_toolchain.cmake file generated still contains `set(CMAKE_GENERATOR_PLATFORM "x64" CACHE STRING "" FORCE)` that prevents Ninja to work.

Conan version = 1.44.1

**Comment by memsharded** [comment]
Hi @julienalofs 

There might be a typo there sorry for that. I think the correct conf is ``-c tools.cmake.cmaketoolchain:generator=Ninja``, could you please try it?

**Comment by julienalofs** [comment]
> Hi @julienalofs
> 
> There might be a typo there sorry for that. I think the correct conf is `-c tools.cmake.cmaketoolchain:generator=Ninja`, could you please try it?

Hi @memsharded

Thanks again for your quick reply.  `-c tools.cmake.cmaketoolchain:generator=Ninja` is working !

Thank you !
